python/ray/dashboard/client/src/pages/dashboard/node-info/NodeInfo.tsx

```diff
@@ -1,40 +1,137 @@
+import { getFnComparator, Order, stableSort } from "../../../common/tableUtils";
 import { sum } from "../../../common/util";
 import { StoreState } from "../../../store";
 import Errors from "./dialogs/errors/Errors";
 import Logs from "./dialogs/logs/Logs";
+import cpuFeature from "./features/CPU";
+import diskFeature from "./features/Disk";
+import makeErrorsFeature from "./features/Errors";
+import gpuFeature from "./features/GPU";
+import gramFeature from "./features/GRAM";
+import hostFeature from "./features/Host";
+import makeLogsFeature from "./features/Logs";
+import ramFeature from "./features/RAM";
+import receivedFeature from "./features/Received";
+import sentFeature from "./features/Sent";
+import {
+  Node,
+  nodeInfoColumnId,
+  NodeInfoFeature,
+  WorkerFeatureData,
+} from "./features/types";
+import uptimeFeature from "./features/Uptime";
+import workersFeature from "./features/Workers";
 import NodeRowGroup from "./NodeRowGroup";
+import { NodeWorkerRow } from "./NodeWorkerRow";
 import TotalRow from "./TotalRow";
 
-const clusterWorkerPids = (
-  rayletInfo: RayletInfoResponse,
-): Map<string, Set<string>> => {
-  // Groups PIDs registered with the raylet by node IP address
-  // This is used to filter out processes belonging to other ray clusters.
-  const nodeMap = new Map();
-  const workerPids = new Set();
-  for (const [nodeIp, { workersStats }] of Object.entries(rayletInfo.nodes)) {
-    for (const worker of workersStats) {
-      if (!worker.isDriver) {
-        workerPids.add(worker.pid.toString());
-      }
+const sortWorkers = (
+  workerFeatureData: WorkerFeatureData[],
+  sortWorkerComparator: any,
+) => {
+  // Sorts idle workers to end, applies the worker comparator function to sort
+  // then returns a new list of worker feature data.
+  const idleSortedClusterWorkers = workerFeatureData.sort((wfd1, wfd2) => {
+    const w1 = wfd1.worker;
+    const w2 = wfd2.worker;
+    if (w2.cmdline[0] === "ray::IDLE") {
+      return -1;
     }
-    nodeMap.set(nodeIp, workerPids);
-  }
-  return nodeMap;
+    if (w1.cmdline[0] === "ray::IDLE") {
+      return 1;
+    }
+    return w1.pid < w2.pid ? -1 : 1;
+  });
+  return sortWorkerComparator
+    ? stableSort(idleSortedClusterWorkers, sortWorkerComparator)
+    : idleSortedClusterWorkers;
+};
+
+const makeGroupedTableContents = (
+  nodes: Node[],
+  sortWorkerComparator: any,
+  sortGroupComparator: any,
+  rayletInfo: RayletInfoResponse | null,
+  nodeInfoFeatures: NodeInfoFeature[],
+) => {
+  const sortedGroups = stableSort(nodes, sortGroupComparator);
+  return sortedGroups.map((node) => {
+    const workerFeatureData: WorkerFeatureData[] = node.workers.map(
+      (worker) => {
+        const rayletWorker =
+          rayletInfo?.nodes?.[node.ip]?.workersStats?.find(
+            (workerStats) => workerStats.pid === worker.pid,
+          ) || null;
+        return {
+          node: node,
+          worker,
+          rayletWorker,
```

Comment:
I think `rayletworker` sounds misleading. Doesn't this just mean the worker stats obtained from raylet? What about something like `workerStat`?

Reply:
That's a fair point, maybe I should rename to rayletWorkerStats. The good thing is, in the v2 of the API, the raylet stats are returned in the same payload as the worker information from the node, so they won't need to be separate arguments.